[AArch64] Change the coercion type of structs with pointer members.

davemgreen opened the pull request:
The aim here is to avoid a ptrtoint->inttoptr round-trip through the function
argument whilst keeping the calling convention the same. Given a struct which
is <= 128bits in size, which can only contain either 1 or 2 pointers, we
convert to a ptr or [2 x ptr] as opposed to the old coercion that uses i64 or
[2 x i64]. This helps alias analysis produce more accurate results.

efriedma-quic commented on clang/lib/CodeGen/Targets/AArch64.cpp:
```diff
@@ -485,6 +485,24 @@ ABIArgInfo AArch64ABIInfo::classifyArgumentType(QualType Ty, bool IsVariadicFn,
+    if (const RecordType *RT = Ty->getAs<RecordType>()) {
+      if (const RecordDecl *RD = RT->getDecl()) {
+        if (all_of(RD->fields(), [](FieldDecl *FD) {
+              return FD->getType()->isPointerOrReferenceType();
```
Correctly collecting all the relevant information is much more complicated than this, in general.  See ABIInfo::isHomogeneousAggregate.  This mostly results in missed optimizations, probably, but I'm concerned this could impact correctness in some C++ corner cases.

You also need to worry about pointers that aren't 64 bits wide.

Please add some testcases for unions. (I don't care if we're conservative for now, just make sure it doesn't explode.)

davemgreen replied:
Thanks - isHomogeneousAggregate was useful to look at. Let me know if there are any other tests that would be useful to add.

efriedma-quic reviewed (commented):
The general idea here makes sense.

efriedma-quic commented on clang/lib/CodeGen/Targets/AArch64.cpp:
```diff
@@ -485,6 +485,39 @@ ABIArgInfo AArch64ABIInfo::classifyArgumentType(QualType Ty, bool IsVariadicFn,
+               ContainsOnlyPointers(FDTy);
+      });
+    };
+    if (getTarget().getTriple().getArch() != llvm::Triple::aarch64_32 &&
```
We also need to handle the __ptr32 qualifier.

I'd prefer to just check that the size of the pointer type itself, rather than trying to derive it from the ABI.

efriedma-quic commented on clang/lib/CodeGen/Targets/AArch64.cpp:
```diff
@@ -485,6 +485,39 @@ ABIArgInfo AArch64ABIInfo::classifyArgumentType(QualType Ty, bool IsVariadicFn,
+             "Expected a 64 or 128bit struct containing pointers");
+      llvm::Type *PtrTy = llvm::PointerType::getUnqual(getVMContext());
+      if (Size == 128)
+        PtrTy = llvm::ArrayType::get(PtrTy, 2);
```
`llvm::ArrayType::get(PtrTy, 2);` is unfortunately not correct if the alignment of the struct is 16 bytes.  (Usually a struct containing only pointers should be 8-byte-aligned, but you can override it with attributes.)

davemgreen replied:
Thanks - I had tested it with an align directive but not realized there was a difference between the "natural alignment" and after it was adjusted.

efriedma-quic commented on clang/lib/CodeGen/Targets/AArch64.cpp:
```diff
@@ -485,6 +485,39 @@ ABIArgInfo AArch64ABIInfo::classifyArgumentType(QualType Ty, bool IsVariadicFn,
+      return all_of(RD->fields(), [&](FieldDecl *FD) {
+        QualType FDTy = FD->getType();
+        if (FDTy->isArrayType())
+          FDTy = QualType(FDTy->getBaseElementTypeUnsafe(), 0);
```
The "unsafe" bit actually matters here because address spaces are qualifiers.  Please use the ASTContext method.

davemgreen commented:
Rebase and ping - thanks.

davemgreen commented on clang/lib/CodeGen/Targets/AArch64.cpp:
```diff
@@ -486,6 +486,39 @@ ABIArgInfo AArch64ABIInfo::classifyArgumentType(QualType Ty, bool IsVariadicFn,
+        return false;
+      const RecordDecl *RD = RT->getDecl();
+      if (const CXXRecordDecl *CXXRD = dyn_cast<CXXRecordDecl>(RD)) {
+        for (const auto &I : CXXRD->bases())
```
I'm not really an expert in clang but I was hoping this would be simpler. I've tried to rewrite the check to be more clear that it is just replacing `i64` with `ptr` where we know the structure contains only pointers. AFAIU even if we get it 'wrong' the ABI remains the same (weirdness around ptrtoint/inttoptr aside). But the idea is this is limited to ptrs.

davemgreen committed:
[AArch64] Change the coercion type of structs with pointer members.

The aim here is to avoid a ptrtoint->inttoptr round-trip throught the function
argument whilst keeping the calling convention the same. Given a struct which
is <= 128bits in size, which can only contain either 1 or 2 pointers, we
convert to a ptr or [2 x ptr] as opposed to the old coercion that uses i64 or
[2 x i64].

davemgreen commented:
Rebase and ping - it feels like this is a decent compromise for keeping the code simple.

efriedma-quic reviewed (approved):
LGTM

eaeltsin commented:
heads up - this change causes `tag-mismatch` under `hwasan` on a bunch of our internal tests.

eaeltsin commented:
The issue seems to be with pointers to functions, adding
```
           FDTy = getContext().getBaseElementType(FDTy);
         return (FDTy->isPointerOrReferenceType() &&
                 getContext().getTypeSize(FDTy) == 64 &&
-                !FDTy->getPointeeType().hasAddressSpace()) ||
+                !FDTy->getPointeeType().hasAddressSpace() &&
+                !FDTy->getPointeeType()->isFunctionType()) ||
                Self(Self, FDTy);
       });
     };
```
to clang/lib/CodeGen/Targets/AArch64.cpp:512 makes the tests pass.

The problematic type in our case seems to be [abseil::FunctionRef](https://github.com/abseil/abseil-cpp/blob/master/absl/functional/function_ref.h#L85)

Please let me know if this is sufficient, reducing internal test to a standalone reproducer might take some time.

eaeltsin commented:
The tests fail with SIGSEGV without hwasan, so there is some problem revealed by this change either in the compiler or in the tests. Bailing out under hwasan is not a fix. 

Disallowing function pointers makes the problem go away, so might be a useful hint. Another hint is that the issue seems to be limited to 128-bit structs only.

We are trying to cook a standalone reproducer.

vitalybuka commented:
> The tests fail with SIGSEGV without hwasan, so there is some problem revealed by this change either in the compiler or in the tests. Bailing out under hwasan is not a fix.
> 
> Disallowing function pointers makes the problem go away, so might be a useful hint. Another hint is that the issue seems to be limited to 128-bit structs only.
> 
> We are trying to cook a standalone reproducer.

@eaeltsin Isn't the original issue are new SIGSEGV on non-hwasan arm builds?

eaeltsin commented:
Sorry I don't have much to add so far. I can reproduce the problem in a pretty complex context (co_await inside the loop, LICM pass seems to make the tests fail), but cannot provide a reduced public reproducer yet.

eaeltsin commented:
https://github.com/llvm/llvm-project/issues/149604 is what @weiguozhi got

Following up my comments in this thread - this specific change looks correct, but we unfortunately have to keep it locally reverted until this lcim/coroutine issue is fixed :(
